pandas/tests/reshape/concat/test_categorical.py

```diff
@@ -124,6 +125,28 @@ def test_categorical_concat_preserve(self):
         )
         tm.assert_frame_equal(res, exp)
 
+    def test_concat_categorical_same_categories_different_order(self):
+        # https://github.com/pandas-dev/pandas/issues/24845
+
+        c1 = CategoricalIndex(["a", "a"], categories=["a", "b"],
+                              ordered=False
+                              )
+        c2 = CategoricalIndex(["b", "b"], categories=["b", "a"],
+                              ordered=False
+                              )
+        c3 = CategoricalIndex(["a", "a", "b", "b"],
+                              categories=["a", "b"], ordered=False
+                              )
+
+        df1 = DataFrame({"A": [1, 2]}, index=c1)
+        df2 = DataFrame({"A": [3, 4]}, index=c2)
+
+        result = concat((df1, df2))
+
+        expected = DataFrame({"A": [1, 2, 3, 4]}, index=c3)
+
+        tm.assert_frame_equal(result, expected)
```

Comment:
It works, but now try `c1.append(c2)` and you should see:
```
CategoricalIndex(['a', 'a', 'a', 'a'], categories=['a', 'b'], ordered=False, dtype='category')
```
which doesn't make sense. It's better you solve the underlying problem.